ELEC-55: STM32F0xx CAN support

tituschow opened the pull request:
This is a big dump of CAN. I will most likely split this into multiple PRs for ease of understanding, but this is here for context.

The idea is that there is a low-level HW layer and a high-level network layer. The HW layer is platform-specific and handles raw CAN messages. The network layer handles device/message IDs, RX/TX queues, RX handlers, and ACK requests.

The idea behind the network layer is that RX/TX queues should be used to buffer CAN messages. RX messages are buffered until an RX FSM is used to process received messages (passing it to a registered handler) or handle ACK responses (finding the corresponding pending ACK request).

TODO: I don't have a good way to determine which messages require ACKs. I think the easiest way to do this is to either reserve a bit for "ACK requested" or build a minimum ID into the codegen.

karlding commented on libraries/ms-common/inc/can_hw.h:
```diff
@@ -0,0 +1,55 @@
+  uint8_t num_filters;
+  bool loopback;
+  CANHwEventHandler handlers[NUM_CAN_HW_EVENTS];
+} CANHwConfig;
```
Any reason why you're using `Config` here as opposed to `Settings`, which seems to be what we've been using for config structs?

tituschow replied:
The other drivers follow a different paradigm (uses an enum to index into hidden storage) while the CAN drivers are more of a catch-all storage/settings struct.

This is mostly because I wrote these parts last year before all of this existed. I'm not entirely sure which approach is better. Thoughts?

karlding commented on libraries/ms-common/src/stm32f0xx/can_hw.c:
```diff
@@ -0,0 +1,172 @@
+  uint8_t tx_status = CAN_Transmit(can_hw->base, &tx_msg);
+  switch (tx_status) {
+    case CAN_TxStatus_NoMailBox:
+    // case CAN_TxStatus_Failed:
```
?

tituschow replied:
TX returns failed on loopback even when it succeeds.

karlding commented on libraries/ms-common/src/stm32f0xx/can_hw.c:
```diff
@@ -0,0 +1,172 @@
+  uint16_t tq = clocks.PCLK_Frequency / 1000 / CAN_HW_PRESCALER / can_hw->bus_speed - 1;
+  uint16_t bs1 = tq * 7 / 8; // 87.5% sample point
+
+  CAN_DeInit(can_hw->base);
```
What's the difference between this line here and when you call `CAN_DeInit(CAN);` at the start of the function?

tituschow replied:
Ah, that was from testing. I was trying to figure out if CAN was being disabled properly. I'll remove the first one.

karlding commented on libraries/ms-common/src/stm32f0xx/can_hw.c:
```diff
@@ -0,0 +1,172 @@
+
+StatusCode can_hw_register_callback(CANHwConfig *can_hw, CANHwEvent event,
+                                    CANHwEventHandlerCb callback, void *context) {
+  if (event > NUM_CAN_HW_EVENTS) {
```
Shouldn't this be ``>=``?

karlding replied:
Should this be ``>=``?

tituschow commented on libraries/ms-common/inc/can_msg.h:
```diff
@@ -0,0 +1,52 @@
+typedef enum {
+  CAN_MSG_TYPE_DATA = 0,
+  CAN_MSG_TYPE_ACK,
+  CAN_MSG_NUM_TYPES
```
Technically we have some that don't follow that convention (`*_mcu`), but I'll change this one.

karlding commented on libraries/ms-common/src/can_ack.c:
```diff
@@ -0,0 +1,116 @@
+// through an array of request pointers to minimize copying
+// ACK requests currently ordered as they were created
+#include "can_ack.h"
+#include "log.h"
```
Is this ``#include`` still necessary?

tituschow replied:
No. I don't really mind leaving `log.h` since it's helpful for debugging, but I'll remove it.

ckitagawa commented on libraries/ms-common/src/can_rx.c:
```diff
@@ -0,0 +1,63 @@
+#include "can_rx.h"
```
Nit: `<stdlib.h>` and `<string.h>` should technically be listed here.

tituschow commented on libraries/ms-common/inc/can_ack.h:
```diff
@@ -0,0 +1,65 @@
+
+// ACK timeout - transit time + computation time
+// Note that this timeout is currently an arbitrary value, but should be minimized.
+#define CAN_ACK_TIMEOUT_MS 10
```
Not really, just didn't have a good reason to. I was thinking it's probably better to have a consistent ACK timeout across all the platforms.

ckitagawa replied:
Fair enough, I suppose this isn't as prone to inconsistent latency as internet based systems.

karlding commented on libraries/ms-common/inc/can_ack.h:
```diff
@@ -0,0 +1,65 @@
+#include "status.h"
+#include "soft_timer.h"
+
+// ACK timeout - transit time + computation time
```
Should this be

```
// ACK timeout = transit time + computation time
              ^^^
```

tituschow replied:
I mean it was meant to be a dash and more of a note than an equation. It's more like "should account for".

karlding commented on libraries/ms-common/src/can.c:
```diff
@@ -0,0 +1,155 @@
+
+    StatusCode result = prv_transmit(can, &tx_msg);
+
+    if (result != STATUS_CODE_OK) {
```
I think we have a macro for this?

```c
status_ok_or_return(result);
```

ckitagawa replied:
Do we want a log statement here and in the `prv_rx_handler`? I can see wanting to be able to debug this if it fails over and over.

tituschow replied:
`can_hw_transmit` returns a statuscode if transmit fails, so that should be enough. A status callback should be good for alerting us about what's gone wrong.

karlding commented on libraries/ms-common/src/can.c:
```diff
@@ -0,0 +1,155 @@
+    CAN_MSG_SET_RAW_ID(&rx_msg, rx_id.raw);
+
+    StatusCode result = can_queue_push(&can->rx_queue, &rx_msg);
+    if (result != STATUS_CODE_OK) {
```
Same here about the macro

```c
status_ok_or_return(result);
```

karlding commented on libraries/ms-common/src/can.c:
```diff
@@ -0,0 +1,155 @@
+      return status_msg(STATUS_CODE_INVALID_ARGS, "CAN: ACK requested for non-critical message");
+    }
+
+    StatusCode ret = STATUS_CODE_OK;
```
Can't you just directly do the assignment?

tituschow replied:
Yeah, it's just more of a really long line thing. It's also why I didn't just wrap the function call in `status_ok_or_return`.

ckitagawa replied:
Personally, I'm okay with the long line as it took me longer to figure out the point of this than the long line would have, but up to you? If you stick with this I would consider not assigning `STATUS_CODE_OK` as the function call is just going to update `ret` and that assignment confused me the most.

karlding commented on libraries/ms-common/src/can_ack.c:
```diff
@@ -0,0 +1,116 @@
+  found_request->response_bitset |= (1 << device);
+
+  if (found_request->callback != NULL) {
+    StatusCode ret = STATUS_CODE_OK;
```
Can't you just initialize this variable with the callback status?

tituschow replied:
Yes, but then the line becomes very long. It compiles down to the same thing.

ckitagawa replied:
I would still drop the `STATUS_CODE_OK` bit...

tituschow commented:
I've addressed the comments and created another ticket (ELEC-202) to track critical messages. ELEC-203 tracks the case where we might want to switch to the indexed method.

karlding commented on libraries/ms-common/src/can_rx_fsm.c:
```diff
@@ -0,0 +1,79 @@
+  }
+
+  fsm_init(rx_fsm, "can_rx_fsm", &can_rx_fsm_handle, can);
+  fsm_state_init(can_rx_fsm_handle, prv_handle_rx);
```
Correct me if I'm wrong, but doesn't not having a return statement in this case result in undefined behaviour if someone checks the status?

tituschow replied:
oops

karlding commented on libraries/ms-common/src/can_ack.c:
```diff
@@ -0,0 +1,113 @@
+  memset(requests, 0, sizeof(*requests));
+
+  objpool_init(&requests->pool, requests->request_nodes, NULL, NULL);
+  requests->num_requests = 0;
```
Same here regarding the `return`?

karlding commented on libraries/ms-common/src/can_queue.c:
```diff
@@ -0,0 +1,21 @@
+#include "can_queue.h"
+
+StatusCode can_queue_init(CANQueue *can_queue) {
+  pqueue_backed_init(&can_queue->pqueue, can_queue->queue_nodes, can_queue->msg_nodes);
```
Same here regarding the `return`?

tituschow commented:
@karlding - Any other concerns?
